arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedDeltaEncodedValuesReader.java

```diff
@@ -0,0 +1,283 @@
+   *
+   * @return the number of values read into output
+   */
+  private int loadMiniBlockToOutput(
+      int remaining, FieldVector vec, int rowId, int typeWidth, IntegerOutputWriter outputWriter)
+      throws IOException {
+
+    // new block; read the block header
+    if (remainingInBlock == 0) {
+      readBlockHeader();
+    }
+
+    // new miniblock, unpack the miniblock
+    if (remainingInMiniBlock == 0) {
+      unpackMiniBlock();
+    }
+
+    // read values from miniblock
+    int valuesReadInMiniBlock = 0;
+    for (int i = miniBlockSizeInValues - remainingInMiniBlock;
+        i < miniBlockSizeInValues && valuesReadInMiniBlock < remaining;
+        i++) {
+      // calculate values from deltas unpacked for current block
+      long outValue = lastValueRead + minDeltaInCurrentBlock + unpackedValuesBuffer[i];
+      lastValueRead = outValue;
+      outputWriter.write(vec, ((long) (rowId + valuesReadInMiniBlock) * typeWidth), outValue);
+      remainingInBlock--;
+      remainingInMiniBlock--;
+      valuesReadInMiniBlock++;
+    }
+
+    return valuesReadInMiniBlock;
+  }
+
+  private void readBlockHeader() {
+    try {
+      minDeltaInCurrentBlock = BytesUtils.readZigZagVarLong(inputStream);
+    } catch (IOException e) {
+      throw new ParquetDecodingException("Can not read min delta in current block", e);
+    }
+    readBitWidthsForMiniBlocks();
+    remainingInBlock = blockSizeInValues;
+    currentMiniBlock = 0;
+    remainingInMiniBlock = 0;
+  }
+
+  /**
+   * mini block has a size of 8*n, unpack 32 value each time
+   *
+   * <p>see org.apache.parquet.column.values.delta.DeltaBinaryPackingValuesReader#unpackMiniBlock
+   */
+  private void unpackMiniBlock() throws IOException {
+    Arrays.fill(this.unpackedValuesBuffer, 0);
+    BytePackerForLong packer =
+        Packer.LITTLE_ENDIAN.newBytePackerForLong(bitWidths[currentMiniBlock]);
+    for (int j = 0; j < miniBlockSizeInValues; j += 8) {
+      ByteBuffer buffer = inputStream.slice(packer.getBitWidth());
+      if (buffer.hasArray()) {
+        packer.unpack8Values(
+            buffer.array(), buffer.arrayOffset() + buffer.position(), unpackedValuesBuffer, j);
+      } else {
+        packer.unpack8Values(buffer, buffer.position(), unpackedValuesBuffer, j);
+      }
+    }
+    remainingInMiniBlock = miniBlockSizeInValues;
+    currentMiniBlock++;
+  }
+
+  // From org.apache.parquet.column.values.delta.DeltaBinaryPackingValuesReader
+  private void readBitWidthsForMiniBlocks() {
+    for (int i = 0; i < miniBlocksPerBlock; i++) {
+      try {
+        bitWidths[i] = BytesUtils.readIntLittleEndianOnOneByte(inputStream);
+      } catch (IOException e) {
+        throw new ParquetDecodingException("Can not decode bitwidth in block header", e);
+      }
+    }
+  }
+
+  /** A functional interface to write long values to into a FieldVector */
+  @FunctionalInterface
+  interface IntegerOutputWriter {
```

Comment:
nit: this Writer writes both int and long value? Shall we call it `IntegralOutPutWriter` instead?